In [2]:
import numpy as np
import pandas as pd
from datetime import datetime
from scipy.stats import norm

In [3]:
# read in all the raw files
first_pilot = pd.read_csv('first_pilot.csv')
second_pilot = pd.read_csv('second_pilot.csv')
control = pd.read_csv('manipulated_data_control_full_dropped.csv')
publicgood = pd.read_csv('manipulated_data_publicgood_full_dropped.csv')
evaluation = pd.read_csv('manipulated_data_evaluation_full_dropped.csv')

In [4]:
first_pilot.head()

_unit_id         _created_at  _golden         _id _missed  \
0  2216121972  3/26/2019 02:26:13    False  4661147698     NaN   
1  2216121972  3/26/2019 02:27:58    False  4661150595     NaN   
2  2216121972  3/26/2019 02:33:59    False  4661159939     NaN   
3  2216121973  3/26/2019 02:26:13    False  4661147690     NaN   
4  2216121973  3/26/2019 02:27:58    False  4661150604     NaN   

          _started_at  _tainted   _channel  _trust  _worker_id _country  \
0  3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
1  3/26/2019 02:27:19     False  clixsense     1.0    38124457      VEN   
2  3/26/2019 02:33:36     False  clixsense     1.0    44386985      TUR   
3  3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
4  3/26/2019 02:27:19     False  clixsense     1.0    38124457      VEN   

  _region      _city             _ip  \
0      68     Ankara      94.54.16.1   
1      19  Agua Fría  190.198.247.68   
2     NaN        NaN   176.88.93.233   
3      68     Ankara      94.54.16.1   
4      19  Agua Fría  190.198.247.68   

   please_select_the_sentiment_feeling_of_the_product_review   1  label  \
0                                                  0         NaN      0   
1                                                  0         NaN      0   
2                                                  0         NaN      0   
3                                                  1         NaN      1   
4                                                  1         NaN      1   

   please_select_the_sentiment_feeling_of_the_product_review_gold  \
0                                                NaN                
1                                                NaN                
2                                                NaN                
3                                                NaN                
4                                                NaN                

                                                text  this_is_a_lion  
0  So there is no way for me to plug it in here i...             NaN  
1  So there is no way for me to plug it in here i...             NaN  
2  So there is no way for me to plug it in here i...             NaN  
3                        Good case, Excellent value.             NaN  
4                        Good case, Excellent value.             NaN

In [7]:
first_pilot.columns

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip',
       'please_select_the_sentiment_feeling_of_the_product_review', '1',
       'label',
       'please_select_the_sentiment_feeling_of_the_product_review_gold',
       'text', 'this_is_a_lion'],
      dtype='object')

In [10]:
first_pilot[first_pilot['_worker_id'] == 44980050].sort_values('_started_at')

_unit_id         _created_at  _golden         _id _missed  \
300  2216130240  3/26/2019 02:25:37     True  4661146706     NaN   
333  2218243929  3/26/2019 02:25:37     True  4661146712     NaN   
362  2218245535  3/26/2019 02:25:37     True  4661146708     NaN   
582  2218246446  3/26/2019 02:25:37     True  4661146711     NaN   
686  2218246733  3/26/2019 02:25:37     True  4661146715     NaN   
0    2216121972  3/26/2019 02:26:13    False  4661147698     NaN   
3    2216121973  3/26/2019 02:26:13    False  4661147690     NaN   
6    2216121974  3/26/2019 02:26:13    False  4661147697     NaN   
9    2216121975  3/26/2019 02:26:13    False  4661147694     NaN   
478  2218245921  3/26/2019 02:26:13     True  4661147692     NaN   

            _started_at  _tainted   _channel  _trust  _worker_id _country  \
300  3/26/2019 02:24:55     False  clixsense     1.0    44980050      TUR   
333  3/26/2019 02:24:55     False  clixsense     1.0    44980050      TUR   
362  3/26/2019 02:24:55     False  clixsense     1.0    44980050      TUR   
582  3/26/2019 02:24:55     False  clixsense     1.0    44980050      TUR   
686  3/26/2019 02:24:55     False  clixsense     1.0    44980050      TUR   
0    3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
3    3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
6    3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
9    3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   
478  3/26/2019 02:25:39     False  clixsense     1.0    44980050      TUR   

    _region   _city         _ip  \
300      68  Ankara  94.54.16.1   
333      68  Ankara  94.54.16.1   
362      68  Ankara  94.54.16.1   
582      68  Ankara  94.54.16.1   
686      68  Ankara  94.54.16.1   
0        68  Ankara  94.54.16.1   
3        68  Ankara  94.54.16.1   
6        68  Ankara  94.54.16.1   
9        68  Ankara  94.54.16.1   
478      68  Ankara  94.54.16.1   

     please_select_the_sentiment_feeling_of_the_product_review   1  label  \
300                                                  1         NaN      1   
333                                                  1         NaN      1   
362                                                  1         NaN      1   
582                                                  1         NaN      1   
686                                                  1         NaN      1   
0                                                    0         NaN      0   
3                                                    1         NaN      1   
6                                                    1         NaN      1   
9                                                    0         NaN      0   
478                                                  1         NaN      1   

     please_select_the_sentiment_feeling_of_the_product_review_gold  \
300                                                1.0                
333                                                1.0                
362                                                1.0                
582                                                1.0                
686                                                1.0                
0                                                  NaN                
3                                                  NaN                
6                                                  NaN                
9                                                  NaN                
478                                                1.0                

                                                  text  this_is_a_lion  
300                                     This is a lion             NaN  
333                                     This is a lion             NaN  
362                                     This is a lion             NaN  
582                                     This is a lion             NaN  
686                          

### Fields 


    _unit_id: A unique ID number created by the system for each question
    _created_at: The time the contributor submitted the judgement 
    _golden: This will be "true" if this is a test question, otherwise it is "false"
    _id: A unique ID number generated for this specific judgment
    _missed: This will be "true" if the row is an incorrect judgment on a test question.
    _started_at: The time at which the contributor started working on the judgement
    _tainted: This will be "true" if the contributor has been flagged for falling below the required accuracy.
    _channel: The work channel that the contributor accessed the job through.
    _trust: The contributor's accuracy. 
    _worker_id: A unique ID number assigned to the contributor.
    _country: The country the contributor is from.
    _region: A region code for the area the contributor is from.
    _city: The city the contributor is from.
    _ip: The IP address for the contributor.
    {{field}}: There will be a column for each field in the job, with a header equal to the field's name.
    {{field}}_gold: The correct answer for the test question.

In general, we're interested in the variance for each question. We can use time of response, trust scores and cities to check the balances between groups. The tainted indicator tells us who was dropped and we can report on their statistics. 

# First pilot

In [3]:
# Check fields
print(first_pilot.columns)

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip',
       'please_select_the_sentiment_feeling_of_the_product_review', '1',
       'label',
       'please_select_the_sentiment_feeling_of_the_product_review_gold',
       'text', 'this_is_a_lion'],
      dtype='object')


### How much time did they take to respond?

In [4]:
first_pilot['_time_to_respond'] = first_pilot._created_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))-first_pilot._started_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))

In [63]:
first_pilot.describe()

_unit_id           _id      _trust    _worker_id  \
count  7.600000e+02  7.600000e+02  760.000000  7.600000e+02   
mean   2.217316e+09  4.661306e+09    0.982455  3.997407e+07   
std    1.054164e+06  9.954427e+04    0.074482  8.022212e+06   
min    2.216122e+09  4.661147e+09    0.500000  6.342275e+06   
25%    2.216122e+09  4.661229e+09    1.000000  3.819853e+07   
50%    2.218246e+09  4.661286e+09    1.000000  4.409445e+07   
75%    2.218246e+09  4.661361e+09    1.000000  4.502098e+07   
max    2.218247e+09  4.661534e+09    1.000000  4.519039e+07   

       please_select_the_sentiment_feeling_of_the_product_review    1  \
count                                         760.000000          0.0   
mean                                            0.810526          NaN   
std                                             0.392143          NaN   
min                                             0.000000          NaN   
25%                                             1.000000          NaN   
50%                                             1.000000          NaN   
75%                                             1.000000          NaN   
max                                             1.000000          NaN   

            label  \
count  760.000000   
mean     0.806579   
std      0.395240   
min      0.000000   
25%      1.000000   
50%      1.000000   
75%      1.000000   
max      1.000000   

       please_select_the_sentiment_feeling_of_the_product_review_gold  \
count                                              460.0                
mean                                                 1.0                
std                                                  0.0                
min                                                  1.0                
25%                                                  1.0                
50%                                                  1.0                
75%                                                  1.0                
max                                                  1.0                

       this_is_a_lion        _time_to_respond  correct_pct  
count             0.0                     760   760.000000  
mean              NaN  0 days 00:01:09.368421    96.447368  
std               NaN  0 days 00:01:09.090237    18.522781  
min               NaN         0 days 00:00:12     0.000000  
25%               NaN  0 days 00:00:32.750000   100.000000  
50%               NaN         0 days 00:00:49   100.000000  
75%               NaN  0 days 00:01:19.250000   100.000000  
max               NaN         0 days 00:07:04   100.000000

In [19]:
first_pilot._country.value_counts()[0:10]/first_pilot.shape[0]

IND    0.164474
UKR    0.078947
VEN    0.065789
ROU    0.052632
PHL    0.052632
RUS    0.052632
TUR    0.052632
EGY    0.039474
IDN    0.039474
NGA    0.039474
Name: _country, dtype: float64

In [7]:
len(first_pilot._worker_id.unique())

77

### For the first pilot, people were right in 93% of the questions, with at least one person being right in any question. People seem to have read the instructions and followed them in 98.16% of the cases (assuming that people who don't read the instructions always answer 'negative' to the 'This is a lion' statement). 

In [8]:
#Accuracy = mean response | answer is one OR (1 - mean response) | answer is 0 (the events are mutually exclussive so you can sum)
first_pilot['correct_pct'] = 100*first_pilot.label*first_pilot.please_select_the_sentiment_feeling_of_the_product_review \
                        + 100*(1-first_pilot.label)*(1-first_pilot.please_select_the_sentiment_feeling_of_the_product_review)
fp_qs = first_pilot.groupby('_unit_id').mean() #Group by question and get the means.

In [9]:
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(fp_qs[fp_qs._golden == False].correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Compliance questions accuracy')
print(fp_qs[fp_qs._golden == True].correct_pct.describe())

Normal questions accuracy
count    100.000000
mean      93.666667
std       14.751971
min       33.333333
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64

Compliance questions accuracy
count     13.000000
mean      98.165283
std        1.522096
min       96.551724
25%       96.969697
50%       97.222222
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64


### The main lesson from our first pilot was that too easy questions would make it hard to detect a treatment effect, since our baseline wouldn't leav much room for improvement. 

In [10]:
fp_sbs_norm = first_pilot[first_pilot._golden == False].groupby('_worker_id').mean() #Group by normal questions and get the means.
fp_sbs_comp = first_pilot[first_pilot._golden == True].groupby('_worker_id').mean() #Group by normal questions and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(fp_sbs_norm.correct_pct.describe())
fp_sbs_norm.correct_pct.hist()
#Accuracy in compliance questions (_golden is True)
print()
print('Compliance questions accuracy')
print(fp_sbs_comp.correct_pct.describe())
fp_sbs_comp.correct_pct.hist()

Normal questions accuracy
count     75.000000
mean      93.666667
std       14.871825
min       25.000000
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64

Compliance questions accuracy
count     77.000000
mean      98.268398
std        7.445914
min       50.000000
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64


In [11]:
#there were a couple subjects who only got test questions
first_pilot[first_pilot._worker_id.isin(list(set(fp_sbs_comp.index).difference(fp_sbs_norm.index)))]

_unit_id         _created_at  _golden         _id _missed  \
399  2218245535  3/26/2019 08:17:08     True  4661533651     NaN   
441  2218245859  3/26/2019 03:03:43     True  4661195697     NaN   
477  2218245859  3/26/2019 08:17:08     True  4661533650     NaN   
482  2218245921  3/26/2019 03:03:43     True  4661195711     NaN   
519  2218246209  3/26/2019 03:03:43     True  4661195716     NaN   
653  2218246531  3/26/2019 08:17:08     True  4661533649     NaN   
685  2218246643  3/26/2019 08:17:08     True  4661533648     NaN   
690  2218246733  3/26/2019 03:03:43     True  4661195702     NaN   
726  2218246811  3/26/2019 03:03:43     True  4661195707     NaN   
759  2218246811  3/26/2019 08:17:08     True  4661533646     NaN   

            _started_at  _tainted    _channel  _trust  _worker_id  \
399  3/26/2019 08:16:34     False   clixsense     1.0    21707202   
441  3/26/2019 03:03:00     False  prizerebel     1.0    44352740   
477  3/26/2019 08:16:34     False   clixsense     1.0    21707202   
482  3/26/2019 03:03:00     False  prizerebel     1.0    44352740   
519  3/26/2019 03:03:00     False  prizerebel     1.0    44352740   
653  3/26/2019 08:16:34     False   clixsense     1.0    21707202   
685  3/26/2019 08:16:34     False   clixsense     1.0    21707202   
690  3/26/2019 03:03:00     False  prizerebel     1.0    44352740   
726  3/26/2019 03:03:00     False  prizerebel     1.0    44352740   
759  3/26/2019 08:16:34     False   clixsense     1.0    21707202   

        ...         _city            _ip  \
399     ...         Paris   37.162.21.43   
441     ...       Chennai  117.207.16.19   
477     ...         Paris   37.162.21.43   
482     ...       Chennai  117.207.16.19   
519     ...       Chennai  117.207.16.19   
653     ...         Paris   37.162.21.43   
685     ...         Paris   37.162.21.43   
690     ...       Chennai  117.207.16.19   
726     ...       Chennai  117.207.16.19   
759     ...         Paris   37.162.21.43   

    please_select_the_sentiment_feeling_of_the_product_review   1  label  \
399                                                  1        NaN      1   
441                                                  1        NaN      1   
477                                                  1        NaN      1   
482                                                  1        NaN      1   
519                                                  1        NaN      1   
653                                                  1        NaN      1   
685                                                  1        NaN      1   
690                                                  1        NaN      1   
726                                                  1        NaN      1   
759                                                  1        NaN      1   

     please_select_the_sentiment_feeling_of_the_product_review_gold  \
399                                                1.0                
441                                                1.0                
477                                                1.0                
482                                                1.0                
519                                                1.0                
653                                                1.0                
685                                                1.0                
690                                                1.0                
726                                                1.0                
759                                                1.0                

               text  this_is_a_lion _time_to_respond  correct_pct  
399  This is a lion             NaN         00:00:34          100  
441  This is a lion             NaN         00:00:43          100  
477  This is a lion             NaN         00:00:34          100  
482  This is a lion             NaN         00:00:43          100  
519  This is a lion             NaN         00:00:43  

### Figure-eight seems to be assigning questions non-randomly, trying that every question is seen which can make some subjects get an unbalanced set of questions. We randomized manually the questions for the experiments in order to correct for this.

### We run a power calculation with the data from the pilot that suggests our test could be underpowered, needing at least 225 subjects in each treatment in order to detect the change in means with near 90% probability. 

In [12]:
def power(effect, n, sigma, alpha = 0.05):
    return norm.cdf(abs(effect)*np.sqrt(n)/(2*sigma) - norm.ppf(1-alpha/2))

In [13]:
#On average every subject answers ten questions, so the minimum impact from the treatment is one more question right or 10% more accuracy.
effect = 100 - 93.666667 #The best we can hope for in our treatment is for 100% accuracy (one more question right on average) given our initial high mean.
sigma = 14.871825 #The standard deviation from the pilot
first_pilot_sample_size = len(fp_sbs_norm.index) #People who answered normal and compliance questions. 
#We'll vary n to see how changing the sample increases the power.
for i in range(1,6):
    n = i*first_pilot_sample_size
    print('Power is ', power(effect,n,sigma), 'for sample size ', n)

Power is  0.45385362684278235 for sample size  75
Power is  0.74147276532363 for sample size  150
Power is  0.8913975965656133 for sample size  225
Power is  0.9580150351716002 for sample size  300
Power is  0.9847455238166528 for sample size  375


# Second pilot

In [14]:
# Check fields
print(second_pilot.columns)

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip', 'animal_present', 'animal_types', 'other_animal_type',
       'animal_present_gold', 'animal_types_gold', 'image_url',
       'other_animal_type_gold'],
      dtype='object')


### This was an unlabeled dataset and that's why some of the columns differ. The analysis will be a tad different but we can check that covariates our fairly similar to the first pilot.

In [15]:
second_pilot['_time_to_respond'] = second_pilot._created_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))-second_pilot._started_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))

In [16]:
second_pilot.describe()

_unit_id           _id      _trust    _worker_id  \
count  5.450000e+02  5.450000e+02  545.000000  5.450000e+02   
mean   2.220961e+09  4.666150e+09    0.714285  4.313023e+07   
std    8.198365e+02  9.146755e+03    0.313202  7.214049e+06   
min    2.220960e+09  4.666137e+09    0.000000  7.344788e+06   
25%    2.220960e+09  4.666145e+09    0.500000  4.451015e+07   
50%    2.220960e+09  4.666149e+09    0.875000  4.515589e+07   
75%    2.220961e+09  4.666152e+09    1.000000  4.516517e+07   
max    2.220962e+09  4.666188e+09    1.000000  4.517808e+07   

       animal_types_gold  other_animal_type_gold        _time_to_respond  
count                0.0                     0.0                     545  
mean                 NaN                     NaN  0 days 00:01:20.137614  
std                  NaN                     NaN  0 days 00:01:14.564848  
min                  NaN                     NaN         0 days 00:00:12  
25%                  NaN                     NaN         0 days 00:00:31  
50%                  NaN                     NaN         0 days 00:00:52  
75%                  NaN                     NaN         0 days 00:01:42  
max                  NaN                     NaN         0 days 00:06:40

In [50]:
#These are the people who got dropped.
second_pilot._tainted.value_counts()

False    535
True      10
Name: _tainted, dtype: int64

In [54]:
#We'll check if their _trust scores are too different
print('People who completed the test - avg. score')
print(second_pilot[second_pilot._tainted == False]._trust.mean())
print('People who completed the test - avg. response time')
print(second_pilot[second_pilot._tainted == False]._time_to_respond.mean())
print('People who were dropped -  avg. score')
print(second_pilot[second_pilot._tainted == True]._trust.mean())
print('People who were dropped -  avg. response time')
print(second_pilot[second_pilot._tainted == True]._time_to_respond.mean())

People who completed the test - avg. score
0.7276364485981306
People who completed the test - avg. response time
0 days 00:01:18.168224
People who were dropped -  avg. score
0.0
People who were dropped -  avg. response time
0 days 00:03:05.500000


### The people who got dropped seem different from those who stayed. We learned here that we should do something about the criteria for dropping people, since it might be related to outcomes. Figure-eight forces us to make test questions, so our solution was to place test questions so easy that no one would fail them (and get dropped). 

In [20]:
second_pilot._country.value_counts()[0:10]/second_pilot.shape[0]

EGY    0.339450
UKR    0.229358
VEN    0.064220
BIH    0.045872
RUS    0.045872
POL    0.036697
LVA    0.036697
SRB    0.036697
ITA    0.036697
BRA    0.036697
Name: _country, dtype: float64

In [18]:
len(second_pilot._worker_id.unique())

34

### The second pilot was a bit less diversified in the number of participants and origins. Howevere, time of completion for each question was 1min and 20 seconds on average, not too different from 1:09 in pilot 1.
Now we'll check accuracy and power of the experiment, but since we had an unlabeled dataset for this pilot (camouflaged images) we'll be making the analysis using a statistic that figure-eight reports on the data called 'Agreement'. Agreement is the in-question variance, or the share of respondents that agreed in the answers of a question. For example, for our first pilot many of the questions would have agreement 1, since most of the people had the correct answer.  

In [40]:
# read in all the raw files
first_pilot_md = pd.read_csv('first_pilot_metadata.csv')
second_pilot_md = pd.read_csv('second_pilot_metadata.csv')

In [42]:
#Check columns
first_pilot_md.columns

Index(['_unit_id', '_golden', '_unit_state', '_trusted_judgments',
       '_last_judgment_at',
       'please_select_the_sentiment_feeling_of_the_product_review',
       'please_select_the_sentiment_feeling_of_the_product_review:confidence',
       '1', 'label',
       'please_select_the_sentiment_feeling_of_the_product_review_gold',
       'text', 'this_is_a_lion'],
      dtype='object')

### Fields 

    _unit_id: A unique ID number created by the system for each row

    _golden: This will be "true" if this is a test question, otherwise it is "false"

    _unit_state: This will be "finalized" if the row has collected all judgments needed, "new" if it has not been launched yet, "judgable" if it requires more judgments, and "golden" if the row is a test question.

    _trusted_judgments: The number of non-tainted judgment the row has accumulated
    
    _last_judgment_at: The time the latest judgment was received

    {{field}}: There will be a column for each field in the job, with a header equal to the fields name

    {{field}}:confidence: If you choose to include confidence values, these columns will be included. They represent the level of agreement between contributors.


In general, we're interested in the variance for each question. We can use time of response, trust scores and cities to check the balances between groups. The tainted indicator tells us who was dropped and we can report on their statistics. 

In [43]:
#Agreement in normal questions (_golden is False)
print('First pilot - Normal questions agreement')
print(first_pilot_md[first_pilot_md._golden == False]['please_select_the_sentiment_feeling_of_the_product_review:confidence'].describe())
#Accuracy in compliance questions (_golden is True)
print()
print('First pilot - Compliance questions agreement')
print(first_pilot_md[first_pilot_md._golden == True]['please_select_the_sentiment_feeling_of_the_product_review:confidence'].describe())
#Agreement in normal questions (_golden is False)
print()
print('Second pilot - Normal questions agreement')
print(second_pilot_md[second_pilot_md._golden == False]['animal_present:confidence'].describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Second pilot - Compliance questions agreement')
print(second_pilot_md[second_pilot_md._golden == True]['animal_present:confidence'].describe())

First pilot - Normal questions agreement
count    100.000000
mean       0.943731
std        0.125110
min        0.647100
25%        1.000000
50%        1.000000
75%        1.000000
max        1.000000
Name: please_select_the_sentiment_feeling_of_the_product_review:confidence, dtype: float64

First pilot - Compliance questions agreement
count    13.000000
mean      0.988431
std       0.013470
min       0.966400
25%       0.979300
50%       1.000000
75%       1.000000
max       1.000000
Name: please_select_the_sentiment_feeling_of_the_product_review:confidence, dtype: float64

Second pilot - Normal questions agreement
count    100.000000
mean       0.908733
std        0.158092
min        0.529400
25%        0.876075
50%        1.000000
75%        1.000000
max        1.000000
Name: animal_present:confidence, dtype: float64

Second pilot - Compliance questions agreement
count    8.000000
mean     0.956625
std      0.043806
min      0.865100
25%      0.948000
50%      0.956250
75%      0.98

### We can see here that agreement in the first pilot was a lot higher, with everyone agreeing on the same (presumably correct) answer from the first quartile onwards. The average agreement was 94% with a standard deviation of 12.5%. Meanwhile, the second pilot seems a lot harder in terms of people agreeing in a single answer: the mean agreement drops to 90.8% and full agreement only becomes visible from the 50% quartile onwards. We know that in the first pilot people were extremely accurate, so more disagreement in the second pilot suggests that the change of database changed as intended; given the picture of a camouflaged animal, some people may see it and others won't, and this is were our treatment of making people take their work more seriously for the good or machine learning or because they're aware that they're being watched could have more room to improve performance scores. We present the new results on statistical power below:

In [47]:
#On average every subject answers ten questions, so the minimum impact from the treatment is one more question right or 10% more accuracy.
effect = 100 - 90.8733 #The best we can hope for in our treatment is for 100% accuracy (one more question right on average) given our initial high mean.
sigma = 15.8092 #The standard deviation from the pilot
first_pilot_sample_size = len(fp_sbs_norm.index) #People who answered normal and compliance questions. 
#We'll vary n to see how changing the sample increases the power.
for i in range(1,6):
    n = i*first_pilot_sample_size
    print('Power is ', power(effect,n,sigma), 'for sample size ', n)

Power is  0.7053434554230589 for sample size  75
Power is  0.9424042017642595 for sample size  150
Power is  0.9911013661795741 for sample size  225
Power is  0.9988156451234848 for sample size  300
Power is  0.9998581516032655 for sample size  375


### A problem with more difficulty and more disagreement, even with the higher variance that this implies, can give you much more power just because of the improvement room for our treatment.

## Summary, during our pilots we learned very important things for our experiments. (i) Depending on when you set the experiment, the distribution of countries may change, but figure-eight seems to make some effort to make them balanced. While this may raise uncertainty on whether subject samples are too different, we can see that they take approximately the same time to answer. (ii) The questions may be allocated non randomly by figure-eight in order to fill quotas of judgements for each question. This introduces a non random element of variance which we try to address by making a manual randomization of the input database. This is also mitigated by the fact that subjects come at random, so even those who only had test questions to answer should not be very different from people who saw more balanced question sets. (iii) A question set that it's too easy will leave little room for improvement, so, we will be using a camouflaged animal set for the experiment. This will introduce some difficulty, and we expect that the instructions to be more careful will make people who got the treatment more attentive and perform better. (iv) the added difficulty meant that some subjects were dropped on performance, which is the very thing we are trying to measure and would be a source of bias. For our experiment, we will be using very easy test questions so no one (hopefully) fails them. 

# Control group

In [55]:
# Check fields
print(control.columns)

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip', 'animal_present', 'animal_present_gold', 'image_url',
       'label'],
      dtype='object')


In [57]:
control['_time_to_respond'] = control._created_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))-control._started_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))

In [58]:
control.describe()

_unit_id           _id       _trust    _worker_id  \
count  1.265000e+03  1.265000e+03  1265.000000  1.265000e+03   
mean   2.228596e+09  4.680842e+09     0.956259  4.353736e+07   
std    1.723026e+06  4.320125e+03     0.154334  3.759920e+06   
min    2.227505e+09  4.680832e+09     0.000000  6.591664e+06   
25%    2.227505e+09  4.680839e+09     1.000000  4.405826e+07   
50%    2.227505e+09  4.680842e+09     1.000000  4.474947e+07   
75%    2.231347e+09  4.680845e+09     1.000000  4.516237e+07   
max    2.231348e+09  4.680866e+09     1.000000  4.520533e+07   

             _time_to_respond  
count                    1265  
mean   0 days 00:00:53.245059  
std    0 days 00:01:05.887886  
min           0 days 00:00:10  
25%           0 days 00:00:23  
50%           0 days 00:00:35  
75%           0 days 00:00:59  
max           0 days 00:12:17

In [66]:
control._country.value_counts()[0:10]/control.shape[0]

VEN    0.683794
EGY    0.071146
COL    0.039526
ESP    0.035573
IND    0.031621
TUR    0.031621
HRV    0.031621
USA    0.015810
ARG    0.011858
CHL    0.007905
Name: _country, dtype: float64

In [68]:
len(control._worker_id.unique())

79

In [77]:
#Preprocessing before getting accuracy numbers
def yes_2_1(x):
    if x == 'yes':
        return 1
    else:
        return 0
    
def label_2_animal(x):
    if x != 'Leaves':
        return 1
    else:
        return 0

control['answer'] = control.animal_present.apply(yes_2_1)
control['true_answer'] = control.label.apply(label_2_animal)

In [79]:
#Accuracy = mean response | answer is one OR (1 - mean response) | answer is 0 (the events are mutually exclussive so you can sum)
control['correct_pct'] = 100*control.true_answer*control.answer \
                        + 100*(1-control.true_answer)*(1-control.answer)
control_qs = control.groupby('_unit_id').mean() #Group by question and get the means.

In [81]:
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(control_qs[control_qs._golden == False].correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Compliance questions accuracy')
print(control_qs[control_qs._golden == True].correct_pct.describe())

Normal questions accuracy
count    232.000000
mean      70.689655
std       35.945331
min        0.000000
25%       33.333333
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64

Compliance questions accuracy
count     16.000000
mean      93.986796
std        3.231171
min       89.189189
25%       91.428571
50%       94.201681
75%       95.317460
max      100.000000
Name: correct_pct, dtype: float64


In [105]:
control_sbs_norm = control[control._golden == False].groupby('_worker_id').mean() #Group by normal questions and get the means.
control_sbs_comp = control[control._golden == True].groupby('_worker_id').mean() #Group by normal questions and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(control_sbs_norm.correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Test questions accuracy')
print(control_sbs_comp.correct_pct.describe())

Normal questions accuracy
count     38.000000
mean      71.982513
std       24.030820
min        0.000000
25%       62.500000
50%       75.000000
75%       87.500000
max      100.000000
Name: correct_pct, dtype: float64

Test questions accuracy
count     79.000000
mean      92.742616
std       24.161196
min        0.000000
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64


In [92]:
#On average every subject answers 6 questions, so the minimum impact from the treatment is one more question right or 16% more accuracy.
effect = 16.6 #this is one more question right
sigma = 24.030820 #The standard deviation from the experiment.
smpl_sz = len(control_sbs_norm.index) #People who answered normal and compliance questions. 
#We'll vary n to see how changing the sample increases the power.
for i in range(1,6):
    n = i*smpl_sz
    print('Power is ', power(effect,n,sigma), 'for sample size ', n)

Power is  0.56716527504949 for sample size  38
Power is  0.8533877995360563 for sample size  76
Power is  0.9579870494294613 for sample size  114
Power is  0.9892272813426868 for sample size  152
Power is  0.9974520303365991 for sample size  190


### Given the conditions of this experiment, we would only find an effect with a probability of 56.7%, but the good news is that a larger study could be able to find an effect statistically different from zero. 

In [93]:
#These are the people who got dropped.
control._tainted.value_counts()

False    1242
True       23
Name: _tainted, dtype: int64

In [94]:
#We'll check if their _trust scores are too different
print('People who completed the test - avg. score')
print(control[control._tainted == False]._trust.mean())
print('People who completed the test - avg. response time')
print(control[control._tainted == False]._time_to_respond.mean())
print('People who were dropped -  avg. score')
print(control[control._tainted == True]._trust.mean())
print('People who were dropped -  avg. response time')
print(control[control._tainted == True]._time_to_respond.mean())

People who completed the test - avg. score
0.9715523349436392
People who completed the test - avg. response time
0 days 00:00:52.763285
People who were dropped -  avg. score
0.13043478260869565
People who were dropped -  avg. response time
0 days 00:01:19.260869


### Dropped people started less trustable, took a few seconds more, but they're so few for our sample that we don't think they will be a concern. 

# Public good group

In [96]:
# Check fields
print(publicgood.columns)

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip', 'animal_present', 'animal_present_gold', 'image_url',
       'label'],
      dtype='object')


In [97]:
publicgood['_time_to_respond'] = publicgood._created_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))-publicgood._started_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))

In [98]:
publicgood.describe()

_unit_id           _id       _trust    _worker_id  \
count  1.515000e+03  1.515000e+03  1515.000000  1.515000e+03   
mean   2.236910e+09  4.693293e+09     0.946890  4.314084e+07   
std    7.931605e+01  5.394808e+03     0.173551  6.057241e+06   
min    2.236910e+09  4.693285e+09     0.000000  1.158291e+07   
25%    2.236910e+09  4.693290e+09     1.000000  4.459232e+07   
50%    2.236910e+09  4.693292e+09     1.000000  4.515890e+07   
75%    2.236910e+09  4.693294e+09     1.000000  4.516983e+07   
max    2.236910e+09  4.693333e+09     1.000000  4.521156e+07   

             _time_to_respond  
count                    1515  
mean   0 days 00:00:58.775577  
std    0 days 00:01:38.485572  
min           0 days 00:00:07  
25%           0 days 00:00:22  
50%           0 days 00:00:33  
75%           0 days 00:00:56  
max           0 days 00:19:38

In [100]:
publicgood._country.value_counts()[0:10]/publicgood.shape[0]

VEN    0.366337
EGY    0.168317
SRB    0.122112
ESP    0.075908
UKR    0.036304
BIH    0.029703
USA    0.029703
TUR    0.029703
IND    0.019802
VNM    0.019802
Name: _country, dtype: float64

In [102]:
len(publicgood._worker_id.unique())

129

In [124]:
publicgood['answer'] = publicgood.animal_present.apply(yes_2_1)
publicgood['true_answer'] = publicgood.label.apply(label_2_animal)
#Accuracy = mean response | answer is one OR (1 - mean response) | answer is 0 (the events are mutually exclussive so you can sum)
publicgood['correct_pct'] = 100*publicgood.true_answer*publicgood.answer \
                        + 100*(1-publicgood.true_answer)*(1-publicgood.answer)
publicgood_qs = publicgood.groupby('_unit_id').mean() #Group by question and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(publicgood_qs[publicgood_qs._golden == False].correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Test questions accuracy')
print(publicgood_qs[publicgood_qs._golden == True].correct_pct.describe())

Normal questions accuracy
count    232.000000
mean      65.373563
std       37.132089
min        0.000000
25%       33.333333
50%       66.666667
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64

Test questions accuracy
count     18.000000
mean      92.274922
std        4.120997
min       82.500000
25%       90.141509
50%       91.987179
75%       95.178571
max      100.000000
Name: correct_pct, dtype: float64


In [125]:
publicgood_sbs_norm = publicgood[publicgood._golden == False].groupby('_worker_id').mean() #Group by normal questions and get the means.
publicgood_sbs_comp = publicgood[publicgood._golden == True].groupby('_worker_id').mean() #Group by normal questions and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(publicgood_sbs_norm.correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Test questions accuracy')
print(publicgood_sbs_comp.correct_pct.describe())

Normal questions accuracy
count     58.000000
mean      65.243499
std       21.579581
min        0.000000
25%       51.250000
50%       65.833333
75%       79.687500
max      100.000000
Name: correct_pct, dtype: float64

Test questions accuracy
count    129.000000
mean      90.715564
std       24.528381
min        0.000000
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64


In [111]:
n = publicgood_sbs_norm.correct_pct.describe()['count']
print('Power is ', power(effect,n,sigma), 'for sample size ', n)

Power is  0.7487133453634108 for sample size  58.0


In [113]:
#These are the people who got dropped.
publicgood._tainted.value_counts()

False    1472
True       43
Name: _tainted, dtype: int64

In [115]:
#We'll check if their _trust scores and times of completion are too different
print('People who completed the test - avg. score')
print(publicgood[publicgood._tainted == False]._trust.mean())
print('People who completed the test - avg. response time')
print(publicgood[publicgood._tainted == False]._time_to_respond.mean())
print('People who were dropped -  avg. score')
print(publicgood[publicgood._tainted == True]._trust.mean())
print('People who were dropped -  avg. response time')
print(publicgood[publicgood._tainted == True]._time_to_respond.mean())

People who completed the test - avg. score
0.9695234374999988
People who completed the test - avg. response time
0 days 00:00:57.084918
People who were dropped -  avg. score
0.17209302325581396
People who were dropped -  avg. response time
0 days 00:01:56.651162


# Evaluation group

In [116]:
# Check fields
print(evaluation.columns)

Index(['_unit_id', '_created_at', '_golden', '_id', '_missed', '_started_at',
       '_tainted', '_channel', '_trust', '_worker_id', '_country', '_region',
       '_city', '_ip', 'animal_present', 'animal_present_gold', 'image_url',
       'label'],
      dtype='object')


In [119]:
evaluation['_time_to_respond'] = evaluation._created_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))-evaluation._started_at.apply(lambda x: datetime.strptime(x, "%m/%d/%Y %H:%M:%S"))
evaluation.describe()

_unit_id           _id       _trust    _worker_id  \
count  1.320000e+03  1.320000e+03  1320.000000  1.320000e+03   
mean   2.237183e+09  4.695386e+09     0.979040  4.093858e+07   
std    8.001923e+01  2.778914e+03     0.132904  1.023581e+07   
min    2.237182e+09  4.695380e+09     0.000000  6.335542e+06   
25%    2.237183e+09  4.695384e+09     1.000000  4.407974e+07   
50%    2.237183e+09  4.695386e+09     1.000000  4.505358e+07   
75%    2.237183e+09  4.695388e+09     1.000000  4.516952e+07   
max    2.237183e+09  4.695404e+09     1.000000  4.521294e+07   

             _time_to_respond  
count                    1320  
mean   0 days 00:00:44.753787  
std    0 days 00:01:00.089994  
min           0 days 00:00:06  
25%           0 days 00:00:18  
50%           0 days 00:00:30  
75%           0 days 00:00:43  
max           0 days 00:11:16

In [120]:
evaluation._country.value_counts()[0:10]/evaluation.shape[0]

VEN    0.344697
EGY    0.106061
SRB    0.106061
ESP    0.068182
VNM    0.064394
UKR    0.060606
GRC    0.041667
COL    0.037879
URY    0.026515
IND    0.022727
Name: _country, dtype: float64

In [121]:
len(evaluation._worker_id.unique())

90

In [123]:
evaluation['answer'] = evaluation.animal_present.apply(yes_2_1)
evaluation['true_answer'] = evaluation.label.apply(label_2_animal)
#Accuracy = mean response | answer is one OR (1 - mean response) | answer is 0 (the events are mutually exclussive so you can sum)
evaluation['correct_pct'] = 100*evaluation.true_answer*evaluation.answer \
                        + 100*(1-evaluation.true_answer)*(1-evaluation.answer)
evaluation_qs = evaluation.groupby('_unit_id').mean() #Group by question and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(evaluation_qs[evaluation_qs._golden == False].correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Test questions accuracy')
print(evaluation_qs[evaluation_qs._golden == True].correct_pct.describe())

Normal questions accuracy
count    232.000000
mean      65.373563
std       36.080908
min        0.000000
25%       33.333333
50%       66.666667
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64

Test questions accuracy
count     18.000000
mean      95.606175
std        2.012094
min       92.592593
25%       94.159664
50%       94.929397
75%       97.121849
max      100.000000
Name: correct_pct, dtype: float64


In [126]:
evaluation_sbs_norm = evaluation[evaluation._golden == False].groupby('_worker_id').mean() #Group by normal questions and get the means.
evaluation_sbs_comp = evaluation[evaluation._golden == True].groupby('_worker_id').mean() #Group by normal questions and get the means.
#Accuracy in normal questions (_golden is False)
print('Normal questions accuracy')
print(evaluation_sbs_norm.correct_pct.describe())
#Accuracy in compliance questions (_golden is True)
print()
print('Test questions accuracy')
print(evaluation_sbs_comp.correct_pct.describe())

Normal questions accuracy
count     48.000000
mean      63.212473
std       23.087103
min        0.000000
25%       50.000000
50%       66.666667
75%       75.000000
max      100.000000
Name: correct_pct, dtype: float64

Test questions accuracy
count     90.000000
mean      94.037037
std       22.304747
min        0.000000
25%      100.000000
50%      100.000000
75%      100.000000
max      100.000000
Name: correct_pct, dtype: float64


In [127]:
n = evaluation_sbs_norm.correct_pct.describe()['count']
print('Power is ', power(effect,n,sigma), 'for sample size ', n)

Power is  0.6674805229149614 for sample size  48.0


In [129]:
#These are the people who got dropped.
evaluation._tainted.value_counts()

False    1298
True       22
Name: _tainted, dtype: int64

In [130]:
#We'll check if their _trust scores and times of completion are too different
print('People who completed the test - avg. score')
print(evaluation[evaluation._tainted == False]._trust.mean())
print('People who completed the test - avg. response time')
print(evaluation[evaluation._tainted == False]._time_to_respond.mean())
print('People who were dropped -  avg. score')
print(evaluation[evaluation._tainted == True]._trust.mean())
print('People who were dropped -  avg. response time')
print(evaluation[evaluation._tainted == True]._time_to_respond.mean())

People who completed the test - avg. score
0.994093220338983
People who completed the test - avg. response time
0 days 00:00:43.292758
People who were dropped -  avg. score
0.09090909090909091
People who were dropped -  avg. response time
0 days 00:02:10.954545
